### PR TITLE
Add rstcheck to CI

### LINKED ISSUE #3622: Add rstcheck for linting rst files on travis
[rstchek](https://github.com/myint/rstcheck) looks like a good tool for helping us to catch some common mistakes on our docs.

More background about this issue here https://github.com/rtfd/readthedocs.org/pull/3498#discussion_r168798352

### PR DESCRIPTION
**Pull request by stsewd**
This is an initial step to have more consistent docs. Closes #3622 

~~Depends on https://github.com/rtfd/readthedocs.org/pull/3498 (since some errors are solved there)~~

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -19,8 +19,10 @@ commands =
     py.test {posargs}
 
 [testenv:docs]
+deps = -r{toxinidir}/requirements/lint.txt
 changedir = {toxinidir}/docs
 commands =
+    rstcheck `find **/*.rst | xargs`
```
Comment (stsewd) [inline tox.ini]: I'm not sure if this is the right place, but looks _logical_ to me, what I don't like is to have to install all the deps from lint.txt. Maybe adding a docs.txt?

Comment (stsewd) [reply tox.ini]: Also the `--recursive` option from rstcheck was added reciently https://github.com/myint/rstcheck/commit/6102bec9a0096e7a855884a8019a0c1fa9f2b8a8, so when that version gets released I will update the command :)

**Comment by stsewd** [comment]
After seen the results from travis I will add some rules to a `.rstcheck.cfg` file

**Comment by stsewd** [comment]
Looks like currently the tool isn't very mature, we can't specify what errors to ignore (there are some very specific cases that are false positives). Anyway, I think this could be a good improvement to merge in the future :). I'm going to see if I can collaborate on rstcheck.

**Comment by ericholscher** [comment]
Big +1 on this, but generally adding linting requires _a lot_ of fixing before merging.

**Comment by stsewd** [comment]
@ericholscher yeah, probably we may want to do separate PRs for fixing some problems, and keep this branch up to date with master.

**Comment by humitos** [comment]
Seems they are not that many:

```
api/bookmarks.rst:5: (INFO/1) Duplicate implicit target name: "".
api/builds.rst:5: (INFO/1) Duplicate implicit target name: "".
api/core.rst:5: (INFO/1) Duplicate implicit target name: "".
api/doc_builder.rst:5: (INFO/1) Duplicate implicit target name: "".
api/projects.rst:5: (INFO/1) Duplicate implicit target name: "".
api/vcs_support.rst:5: (INFO/1) Duplicate implicit target name: "".
business/index.rst:2: (INFO/1) Duplicate explicit target name: "readthedocs.com".
custom_installs/local_rtd_vm.rst:2: (INFO/1) Enumerated list start value not ordinal-1: "2" (ordinal 2)
dmca/index.rst:4: (INFO/1) Hyperlink target "our process" is not referenced.
```

I think we can fix them in this PR also, right?

File: tox.ini (hunk 1 above)
Comment (humitos) [reply tox.ini]: It's already released, `3.2`.

I also noted that you made a PR there! Hehe! Nice job! :)

**Comment by stsewd** [comment]
@humitos the first errors `target name: ""` are false positives, I'm [already collaborating](https://github.com/myint/rstcheck/pull/43) to `rstcheck` so we can used in a more configurable way :)

File: tox.ini [hunk 2]
```diff
@@ -19,8 +19,10 @@ commands =
     py.test {posargs}
 
 [testenv:docs]
+deps = -r{toxinidir}/requirements/lint.txt
 changedir = {toxinidir}/docs
 commands =
+    bash -c 'rstcheck `find **/*.rst | xargs`'
```
Comment (humitos) [inline tox.ini]: I think we should create a `[testenv:doclint]` with the rstcheck linting process instead of making it part of the `docs` building

File: tox.ini [hunk 3]
```diff
@@ -19,11 +19,15 @@ commands =
     py.test {posargs}
 
 [testenv:docs]
+changedir = {toxinidir}/docs
+commands =
+    sphinx-build -b html -d {envtmpdir}/doctrees .  {envtmpdir}/html
+
+[testenv:docs-lint]
 deps = -r{toxinidir}/requirements/lint.txt
```
Comment (humitos) [inline tox.ini]: Here. You just need `rstcheck` instead all the linting packages, I'd say

(or maybe creating another file `docs-lint.txt`?)

Comment (stsewd) [reply tox.ini]: We need also sphinx for more integration https://github.com/myint/rstcheck#sphinx

Comment (humitos) [reply tox.ini]: Nice! So, I'd say that's better to have a separate reqs file.

File: requirements/docs-lint.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+-r pip.txt
```
Comment (stsewd) [inline requirements/docs-lint.txt]: Not sure if we need all the packages from pip.txt, but just to follow similar convention, and we need Sphinx installed.

Comment (humitos) [reply requirements/docs-lint.txt]: No we don't need them. That's the idea of using a different file.

Add here just the packages that are needed. As far as I know, `sphinx` and `rstcheck` (also, do not pin them)

Comment (humitos) [reply requirements/docs-lint.txt]: No we don't need them. That's the idea of using a different file.

Add here just the packages that are needed. As far as I know, `sphinx` and `rstcheck` (also, do not pin them)

File: .travis.yml [hunk 5]
```diff
@@ -7,6 +7,8 @@ matrix:
   include:
     - python: 2.7
       env: TOXENV=docs
+    - python: 2.7
```
Comment (humitos) [inline .travis.yml]: Py3? :)

**Comment by stsewd** [comment]
The 3.3 version of rstcheck was released! https://github.com/myint/rstcheck#id1

Soon I'll be updating this PR :)

**Comment by stsewd** [comment]
Done, I made sure that all the links were shown correctly and that the deletes were not used in other parts of the docs.

**Comment by stsewd** [comment]
Also, sorry for so many commits, you can squashed into one if you want

**Comment by stsewd** [comment]
I missed one file, now the linter for docs should pass.

File: docs/.rstcheck.cfg [hunk 6]
```diff
@@ -0,0 +1,4 @@
+[rstcheck]
+ignore_directives=automodule,http:get
+ignore_roles=djangosetting,setting
+ignore_messages=(Duplicate implicit target name: ".*")|(Hyperlink target ".*" is not referenced)
```
Comment (stsewd) [inline docs/.rstcheck.cfg]: Those errors are from some hyperlinks that are referenced globally. The duplicates are really false positives.

File: docs/dmca/index.rst [hunk 7]
```diff
@@ -4,32 +4,36 @@ DMCA Takedown Policy
 These are the guidelines that Read the Docs follows when handling DMCA takedown
 requests and takedown counter requests. If you are a copyright holder wishing to
 submit a takedown request, or an author that has been notified of a takedown
-request, please familiarize yourself with `our process <Takedown Process_>`_.
+request, please familiarize yourself with `our process`__.
```
Comment (humitos) [inline docs/dmca/index.rst]: Why we don't leave the extended name here? Isn't it clearer to understand where this will go?

Comment (humitos) [reply docs/dmca/index.rst]: (the same for the other changes in this file)

File: docs/.rstcheck.cfg (hunk 6 above)
Comment (humitos) [reply docs/.rstcheck.cfg]: By ignoring this now that are false positivies, we won't detect the ones that are true when we updated the docs :(

Comment (stsewd) [reply docs/.rstcheck.cfg]: Yeah, I going to try to solve others and see if I can get a better regex here

File: docs/dmca/index.rst (hunk 7 above)
Comment (stsewd) [reply docs/dmca/index.rst]: rstcheck didn't recognize those and marked it as errors. And I didn't find that syntax on the quickref http://docutils.sourceforge.net/docs/user/rst/quickref.html#internal-hyperlink-targets.

File: docs/.rstcheck.cfg (hunk 6 above)
Comment (stsewd) [reply docs/.rstcheck.cfg]: There is a better way to expose global hiperlinks? Maybe something like this https://stackoverflow.com/a/19543591/5689214? And for the duplicates target name the most are from subheaders with the same name.

**Comment by stsewd** [comment]
> Optimization: is it possible to trigger this docs-lint env _only when there are changes in the docs/ directory?

I'm not sure if that would be possible, maybe a setting on travis? Since the linting is very quick, the heavy stuff comes from travis and installing the deps p:

File: docs/dmca/index.rst (hunk 7 above)
Comment (humitos) [reply docs/dmca/index.rst]: `our process <Takedown Process_>`_

links to the "Takedown Process" title (L20 in `dcma/index.rst`)

If you remove the title from the link, like this

    `our process`__

where that link will follow you? If you build the docs locally, you will find that link doesn't work.

File: docs/.rstcheck.cfg (hunk 6 above)
Comment (humitos) [reply docs/.rstcheck.cfg]: Why not use the `:ref:` directive for linking across files?

http://www.sphinx-doc.org/es/stable/markup/inline.html#role-ref

File: docs/automatic-redirects.rst [hunk 8]
```diff
@@ -33,7 +33,7 @@ Supported Top-Level Redirects
 .. note:: These "implicit" redirects are supported for legacy reasons.
           We will not be adding support for any more magic redirects.
           If you want additional redirects,
-          they should live at a prefix like :ref:`page-redirect`
+          they should live at a prefix like `Redirecting to a Page`_
```
Comment (humitos) [inline docs/automatic-redirects.rst]: Why did you change these ones?

> Using ref is advised over standard reStructuredText links to sections (like `Section title`_) because it works across files, when section headings are changed, and for all builders that support cross-references.

from http://www.sphinx-doc.org/es/stable/markup/inline.html#role-ref

File: docs/contribute.rst [hunk 9]
```diff
@@ -1,5 +1,3 @@
-.. _contributing-to-read-the-docs:
```
Comment (humitos) [inline docs/contribute.rst]: Be careful when removing these ones. This one in particular, is used at https://github.com/rtfd/readthedocs.org/pull/3624/files#diff-c7f13470cacdf0634e185d4313d02a12R55 (L57 in `design.rst`) so that link is probably dead after removing this.

File: docs/localization.rst [hunk 10]
```diff
@@ -1,5 +1,3 @@
-.. _`Localization of Documentation`:
```
Comment (humitos) [inline docs/localization.rst]: This one is used in `faq.rst`

File: docs/.rstcheck.cfg (hunk 6 above)
Comment (stsewd) [reply docs/.rstcheck.cfg]: rstcheck doesn't search for global refs and got marked as unused. But using ref is fine.

File: docs/automatic-redirects.rst (hunk 8 above)
Comment (stsewd) [reply docs/automatic-redirects.rst]: mmm, well I don't like the idea of exposing a lot of links as global refs and more when they aren't used on the global scope, but I will revert those if you consider.

File: docs/contribute.rst (hunk 9 above)
Comment (stsewd) [reply docs/contribute.rst]: Ups, I just missed that one. I usually do a grep to search those references.

File: docs/localization.rst (hunk 10 above)
Comment (stsewd) [reply docs/localization.rst]: Yeah, and was changed to use the `doc` directive as rtd recommends

https://github.com/rtfd/readthedocs.org/pull/3624/files#diff-c6f0a0722c6a2f86277535d7bcec7f8cR167

**Comment by stsewd** [comment]
@humitos done, just one link was broken, the others were properly modified. By the way I use `sphinx-autobuild` to building the docs on dev.

File: docs/.rstcheck.cfg (hunk 6 above)
Comment (stsewd) [reply docs/.rstcheck.cfg]: And also I change others place to use the `doc` directive https://docs.readthedocs.io/en/latest/docs.html#content.

File: docs/dmca/index.rst (hunk 7 above)
Comment (stsewd) [reply docs/dmca/index.rst]: My mouse is on `our process` but it doesn't appear on the screeenshot :grin:

**Comment by stsewd** [comment]
I just update this branch to the current master and fix an issue with some docs from upstream.

We only need to agree on using `:ref:` role for internal referencing or use the auto generated references from the titles.

**Comment by ericholscher** [comment]
I don't have a strong opinion on using `:ref:` everywhere -- it is annoying to have to add a target to reference things inside the same document. 

What more is needed here? I'd like to get this running in prod, and it looks it's mostly good to go.

**Comment by stsewd** [comment]
I just rebased and solved others conflicts from upstream, this should be ready to go, also I checked that all links work properly.

**Comment by stsewd** [comment]
Also, I found this target on the docs` Makefile `make livehtml`

**Comment by ericholscher** [comment]
@stsewd yea, that lets you auto-rebuild the docs if you have `sphinx-autobuild` installed.

File: docs/automatic-redirects.rst (hunk 8 above)
Comment (ericholscher) [reply docs/automatic-redirects.rst]: I wonder if we can make a Sphinx extension that automatically ads each section heading as a global ref, prefixed by pagename or similar? That way you wouldn't need to declare all the refs, but still be able to reference them. That's definitely for another ticket/PR though.

Comment (ericholscher) [reply docs/automatic-redirects.rst]: http://www.sphinx-doc.org/en/stable/markup/inline.html#role-numref looks interesting, via https://github.com/sphinx-doc/sphinx/pull/2963

Comment (ericholscher) [reply docs/automatic-redirects.rst]: Hmm, looks like those are still for _explicit_ section refs, not implicitly creating them, ignore me :)

Comment (ericholscher) [reply docs/automatic-redirects.rst]: Found it: http://www.sphinx-doc.org/en/stable/ext/autosectionlabel.html -- it's already built into Sphinx :)

Comment (humitos) [reply docs/automatic-redirects.rst]: Like it!

**Comment by humitos** [comment]
Goooood work! :)
